### PR TITLE
DC-2845: there should always be at least one message in a conversation

### PR DESCRIPTION
**Pull request by rwalpole**

### CODE REVIEW

**Review by matthewgroom — DISMISSED** [review]

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala [hunk 1]
```diff
@@ -46,22 +48,16 @@ object ConversationMetaData {
     )
   }
 
-  private def findLatestMessage(coreConversation: Conversation): Option[Message] =
-    coreConversation.messages.sortWith(_.created.getMillis > _.created.getMillis).headOption match {
-      case Some(message) => Some(message)
-      case _             => None
-    }
+  private def findLatestMessage(coreConversation: Conversation): Message =
+    coreConversation.messages.sortBy(_.created.getMillis).reverse.head
```
Comment (muralishub) [inline app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:52]: can we use headOption ? we may atleast have a message for a conversation , but still I feel its better have headOption

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala [hunk 2]
```diff
@@ -72,18 +68,15 @@ object ConversationMetaData {
 
   private def findUnreadMessagesByParticipant(participantId: Int, coreConversation: Conversation): Boolean = {
     val messages = coreConversation.messages
-    val matchingIds = coreConversation.messages
-      .flatMap(_.readBy)
-      .map(_.id)
-      .count(int => int === participantId)
-    matchingIds =!= messages.size
+    val messagesReadByParticipant = messages.filter(_.readBy.map(_.id).contains(participantId))
```
Comment (muralishub) [inline app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:71]: Name of function is find but it returns a  boolean

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala (hunk 1 above)
Comment (muralishub) [reply app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:52]: May be its safe becuase its nonempty list ?

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala [hunk 3]
```diff
@@ -56,8 +57,14 @@ class ConversationRepository @Inject()(implicit connector: MongoConnector)
           Future.successful(false)
       }
 
-  def getConversations(enrolment: Enrolment)(implicit ec: ExecutionContext): Future[List[Conversation]] =
-    find(findByEnrolmentQuery(enrolment): _*)
+  def getConversations(enrolment: Enrolment)(implicit ec: ExecutionContext): Future[List[Conversation]] = {
+    import uk.gov.hmrc.securemessage.models.core.Conversation.conversationFormat
```
Comment (muralishub) [inline app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:61]: can we use this top level ? so we  might need this for another function ?

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ApiConversation.scala [hunk 4]
```diff
@@ -16,6 +16,7 @@
 
 package uk.gov.hmrc.securemessage.controllers.models.generic
 
+import cats.data.NonEmptyList
```
Comment (elabeca) [inline app/uk/gov/hmrc/securemessage/controllers/models/generic/ApiConversation.scala:19]: ❤️

**Review by elabeca — APPROVED** [review]

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala (hunk 2 above)
Comment (elabeca) [reply app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:71]: I think that's a merge issue, I thought @matthewgroom fixed that name?

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala (hunk 1 above)
Comment (muralishub) [inline app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:52]: Even if its nonEmptyList is it still safe to use head ?

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala (hunk 2 above)
Comment (matthewgroom) [reply app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:71]: That's my fault Murali. I thought I changed it to anyUnreadMessages. Sorry @rwalpole

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala (hunk 1 above)
Comment (rwalpole) [reply app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:52]: That is the whole point of using NEL @muralishub - the head is no longer an `Option`

Comment (muralishub) [reply app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:52]: Thanks

**Review by elabeca — APPROVED** [review]

File: app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala (hunk 2 above)
Comment (rwalpole) [reply app/uk/gov/hmrc/securemessage/controllers/models/generic/ConversationMetadata.scala:71]: Updated.

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala (hunk 3 above)
Comment (rwalpole) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:62]: Moved.

**Pull request merged**